cdap-metrics-sidecar/src/main/java/io/cdap/cdap/jmx/metrics/JMXMetricsCollector.java

```diff
@@ -0,0 +1,164 @@
+import io.cdap.cdap.api.metrics.MetricsCollectionService;
+import io.cdap.cdap.api.metrics.MetricsCollector;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.proto.id.NamespaceId;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryMXBean;
+import java.lang.management.MemoryUsage;
+import java.lang.management.ThreadMXBean;
+import java.net.MalformedURLException;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.management.MBeanServerConnection;
+import javax.management.remote.JMXConnector;
+import javax.management.remote.JMXConnectorFactory;
+import javax.management.remote.JMXServiceURL;
+
+/**
+ * Represents a service  that runs along with other services to collect resource usage metrics and publish them to
+ * {@link MetricsCollectionService}. For this service to work, the jvm process needs to expose JMX server on the same
+ * port that this service polls. To do this, the following JAVA OPTS need to be set:
+ * {@code -Dcom.sun.management.jmxremote=true -Dcom.sun.management.jmxremote.port=11022
+ * -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.authenticate=false}
+ * This service also requires {@code SERVICE_NAME} and {@code JMX_SERVER_PORT} env variables to be set which provide
+ * the component tag for setting metrics context and port to connect to jmx server on localhost respectively.
+ */
+public class JMXMetricsCollector extends AbstractScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(JMXMetricsCollector.class);
+  private static final long MEGA_BYTE = 1024 * 1024, MAX_PORT = (1 << 16) - 1;
+  private static final String SERVICE_URL_FORMAT = "service:jmx:rmi:///jndi/rmi://%s:%s/jmxrmi";
+  private final String componentName;
+  private final CConfiguration cConf;
+  private final MetricsCollectionService metricsCollectionService;
+  private final JMXServiceURL serviceUrl;
+  private final Environment env;
+  private ScheduledExecutorService executor;
+
+  @Inject
+  public JMXMetricsCollector(CConfiguration cConf, MetricsCollectionService metricsCollectionService, Environment env)
+    throws MalformedURLException {
+    this.cConf = cConf;
+    int serverPort = cConf.getInt(Constants.JMXMetricsCollector.SERVER_PORT);
+    if (serverPort < 0 || serverPort > MAX_PORT) {
+      throw new IllegalArgumentException(
+        String.format("%s variable (%d) is not a valid port number.",
+                      Constants.JMXMetricsCollector.SERVER_PORT, serverPort));
+    }
+    String serverUrl = String.format(SERVICE_URL_FORMAT, "localhost", serverPort);
+    componentName = env.getVariable(Constants.JMXMetricsCollector.COMPONENT_NAME_ENV_VAR);
```

Comment:
nit: Use "this.componentName" to follow convention.

Reply:
Done.